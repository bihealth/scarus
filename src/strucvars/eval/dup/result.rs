//! Data structures for representing the actual results of DEL/copy number gain.

use crate::strucvars::{
    data::{clingen_dosage, hgnc::GeneIdInfo},
    eval::{
        common::{FunctionalElement, GeneOverlap, ScoreRange, SuggestedScore},
        result::{ClinvarSvOverlap, GnomadSvOverlap, Pvs1Result},
    },
};

/// Evaluation results for each section of the ACMG rule.
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
pub enum Section {
    /// Results of Section G1.
    G1(G1),
    /// Results of Section G2.
    G2(G2),
    /// Results of Section G3.
    G3(G3),
    /// Results of Section G4.
    G4(G4),
}

impl SuggestedScore for Section {
    fn suggested_score(&self) -> f32 {
        match self {
            Section::G1(G1::G1A(g1a)) => g1a.suggested_score(),
            Section::G1(G1::G1B(g1b)) => g1b.suggested_score(),
            Section::G2(G2::G2A(g2a)) => g2a.suggested_score(),
            Section::G2(G2::G2B(g2b)) => g2b.suggested_score(),
            Section::G2(G2::G2C(g2c)) => g2c.suggested_score(),
            Section::G2(G2::G2D(g2d)) => g2d.suggested_score(),
            Section::G2(G2::G2E(g2e)) => g2e.suggested_score(),
            Section::G2(G2::G2F(g2f)) => g2f.suggested_score(),
            Section::G2(G2::G2G(g2g)) => g2g.suggested_score(),
            Section::G2(G2::G2H(g2h)) => g2h.suggested_score(),
            Section::G2(G2::G2I(g2i)) => g2i.suggested_score(),
            Section::G2(G2::G2J(g2j)) => g2j.suggested_score(),
            Section::G2(G2::G2L(g2l)) => g2l.suggested_score(),
            Section::G3(G3::G3A(g3a)) => g3a.suggested_score(),
            Section::G3(G3::G3B(g3b)) => g3b.suggested_score(),
            Section::G3(G3::G3C(g3c)) => g3c.suggested_score(),
            Section::G4(G4::G4Patho(g4p)) => g4p.suggested_score(),
            Section::G4(G4::G4N(g4n)) => g4n.suggested_score(),
            Section::G4(G4::G4O(g4o)) => g4o.suggested_score(),
        }
    }
}

impl ScoreRange for Section {
    fn min_score(&self) -> f32 {
        match self {
            Section::G1(G1::G1A(_)) => 0.0,
            Section::G1(G1::G1B(_)) => -0.6,
            Section::G2(G2::G2A(_)) => 1.0,
            Section::G2(G2::G2B(_)) => 0.0,
            Section::G2(G2::G2C(_)) => -1.0,
            Section::G2(G2::G2D(_)) => -1.0,
            Section::G2(G2::G2E(_)) => 0.0,
            Section::G2(G2::G2F(_)) => -1.0,
            Section::G2(G2::G2G(_)) => 0.0,
            Section::G2(G2::G2H(_)) => 0.0,
            Section::G2(G2::G2I(_)) => 0.0,
            Section::G2(G2::G2J(_)) => 0.0,
            Section::G2(G2::G2L(_)) => 0.0,
            Section::G3(G3::G3A(_)) => 0.0,
            Section::G3(G3::G3B(_)) => 0.45,
            Section::G3(G3::G3C(_)) => 0.9,
            Section::G4(G4::G4Patho(_)) => -0.9,
            Section::G4(G4::G4N(_)) => -0.9,
            Section::G4(G4::G4O(_)) => -1.0,
        }
    }

    fn max_score(&self) -> f32 {
        match self {
            Section::G1(G1::G1A(_)) => 0.0,
            Section::G1(G1::G1B(_)) => -0.6,
            Section::G2(G2::G2A(_)) => 1.0,
            Section::G2(G2::G2B(_)) => 0.0,
            Section::G2(G2::G2C(_)) => -1.0,
            Section::G2(G2::G2D(_)) => -1.0,
            Section::G2(G2::G2E(_)) => 0.0,
            Section::G2(G2::G2F(_)) => 0.0,
            Section::G2(G2::G2G(_)) => 0.0,
            Section::G2(G2::G2H(_)) => 0.0,
            Section::G2(G2::G2I(_)) => 0.9,
            Section::G2(G2::G2J(_)) => 0.0,
            Section::G2(G2::G2L(_)) => 0.0,
            Section::G3(G3::G3A(_)) => 0.0,
            Section::G3(G3::G3B(_)) => 0.45,
            Section::G3(G3::G3C(_)) => 0.9,
            Section::G4(G4::G4Patho(_)) => 0.9,
            Section::G4(G4::G4N(_)) => 0.9,
            Section::G4(G4::G4O(_)) => 0.0,
        }
    }
}

/// Enumeration of the categories for the copy number gain evaluation, Section 1.
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
pub enum G1 {
    /// Contains protein-coding or other known functionally important elements.
    G1A(G1A),
    /// Does NOT contain protein-coding or any functionally important elements.
    G1B(G1B),
}

/// Result of the G1A subsection (important feature).
#[derive(Debug, Clone, PartialEq, Default, serde::Deserialize, serde::Serialize)]
pub struct G1A {
    /// Overlapping transcripts/genes.
    pub genes: Vec<GeneOverlap>,
    /// Overlapping functional elements.
    pub functional_elements: Vec<FunctionalElement>,
}

impl SuggestedScore for G1A {
    fn suggested_score(&self) -> f32 {
        0.0
    }
}

/// Result of the G1B subsection (no important feature).
#[derive(Debug, Clone, PartialEq, Default, serde::Deserialize, serde::Serialize)]
pub struct G1B {
    // no members
}

impl SuggestedScore for G1B {
    fn suggested_score(&self) -> f32 {
        -0.6
    }
}

/// Enumeration of the categories for the structural variant evaluation, Section 2.
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
pub enum G2 {
    /// Complete overlap; Tthe TS gene or minimal criticala reegion is fully contained
    /// within the observed copy number gain.
    G2A(G2A),
    /// Partial overlap of an established TS region.
    G2B(G2B),
    /// Identical in gene content to the established benign copy number gain.
    G2C(G2C),
    /// Smaller than established benign copy number gain, breakpoint(s) does not
    /// interrupt protein-coding genes.
    G2D(G2D),
    /// Smaller than established benign copy number gain, breakpoint(s) potentially
    /// interrupt protein-coding genes.
    G2E(G2E),
    /// Larger tahn known benign copy-number gain, does not include additional
    /// protein-coding genes.
    G2F(G2F),
    /// Overlaps a benign copy number gain but includes additional material.
    G2G(G2G),
    /// TS gene fully contained within obsrved copy-number gain.
    G2H(G2H),
    /// Both breakpoints are within the asme gene (gene-level sequence variant, possibly
    /// resulting in LoF.
    G2I(G2I),
    /// One breakpoint is within an established TS gene, patient's phenotype is either
    /// inconsistent with what is expecte for LOF or that gene OR unknown
    G2J(G2J),
    /// One or both breakpoints are within gene(s) of no established clinical significance.
    G2L(G2L),
}

/// Result of the G2A subsection (complete overlap).
#[derive(Debug, Clone, PartialEq, Default, serde::Deserialize, serde::Serialize)]
pub struct G2A {
    /// Suggested score for the subsection.
    pub suggested_score: f32,
    /// Overlapping TS genes.
    pub ts_genes: Vec<clingen_dosage::Gene>,
    /// Overlapping TS genomic regions.
    pub ts_regions: Vec<clingen_dosage::Region>,
}

impl SuggestedScore for G2A {
    fn suggested_score(&self) -> f32 {
        self.suggested_score
    }
}

/// Result of the G2B subsection (partial overlap of established TS region).
#[derive(Debug, Clone, PartialEq, Default, serde::Deserialize, serde::Serialize)]
pub struct G2B {
    /// Overlapping TS genomic regions.
    pub ts_regions: Vec<clingen_dosage::Region>,
}

impl SuggestedScore for G2B {
    fn suggested_score(&self) -> f32 {
        0.0
    }
}

/// Result of the G2C subsection (identical in gene content to the established
/// benign TS region).
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
pub struct G2C {
    /// Suggested score for the subsection.
    pub suggested_score: f32,
    /// Overlapping TS genomic regions.
    pub benign_regions: Vec<clingen_dosage::Region>,
}

impl SuggestedScore for G2C {
    fn suggested_score(&self) -> f32 {
        self.suggested_score
    }
}

/// Result of the G2D subsection (smaller than established benign copy-number gain,
/// breakpoint(s) does not interrupt protein-coding gene).
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
pub struct G2D {
    /// Suggested score for the subsection.
    pub suggested_score: f32,
    /// Overlapping benign genomic region.
    pub benign_region: clingen_dosage::Region,
}

impl SuggestedScore for G2D {
    fn suggested_score(&self) -> f32 {
        self.suggested_score
    }
}

/// Result of the G2E subsection (smaller than established benign copy-number gain,
/// breakpoint(s) potentially interrupt protein-coding gene).
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
pub struct G2E {
    /// Suggested score for the subsection.
    pub suggested_score: f32,
    /// Overlapping benign genomic regions.
    pub benign_region: clingen_dosage::Region,
    /// Potentially affected genes.
    pub genes: Vec<GeneIdInfo>,
    /// Overlapping functional elements.
    pub functional_elements: Vec<FunctionalElement>,
}

impl SuggestedScore for G2E {
    fn suggested_score(&self) -> f32 {
        self.suggested_score
    }
}

/// Result of G2F subsection (larger than known benign copy-number gain,
/// does not include additional protein-coding genes).
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
pub struct G2F {
    /// Suggested score for the subsection.
    pub suggested_score: f32,
    /// Overlapping benign genomic regions.
    pub benign_region: clingen_dosage::Region,
}

impl SuggestedScore for G2F {
    fn suggested_score(&self) -> f32 {
        self.suggested_score
    }
}

/// Result of the G2G subsection (overlaps a benign copy-number gain but includes
/// additional genomic material).
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
pub struct G2G {
    /// Suggested score for the subsection.
    pub suggested_score: f32,
    /// Overlapping benign genomic regions.
    pub benign_region: clingen_dosage::Region,
    /// Additional genes.
    pub genes: Vec<GeneIdInfo>,
    /// Overlapping functional elements.
    pub functional_elements: Vec<FunctionalElement>,
}

impl SuggestedScore for G2G {
    fn suggested_score(&self) -> f32 {
        self.suggested_score
    }
}

/// Result of the G2H subsection (HI gene fully contained within observed copy-number
/// gain).
#[derive(Debug, Clone, PartialEq, Default, serde::Deserialize, serde::Serialize)]
pub struct G2H {
    /// Suggested score for the subsection.
    pub suggested_score: f32,
    /// Overlapping HI gene information.
    pub hi_genes: Vec<GeneIdInfo>,
}

impl SuggestedScore for G2H {
    fn suggested_score(&self) -> f32 {
        self.suggested_score
    }
}

/// Result of the G2I subsection (both breakpoint are within the same gene, potential
/// LOF, fall back to PVS1).
#[derive(Debug, Clone, PartialEq, Default, serde::Deserialize, serde::Serialize)]
pub struct G2I {
    /// Suggested score for the subsection.
    pub suggested_score: f32,
    /// PVS assessment result.
    pub pvs1_result: Pvs1Result,
    /// HI genes for which the assessment was made.
    pub hi_genes: Vec<GeneIdInfo>,
}

impl SuggestedScore for G2I {
    fn suggested_score(&self) -> f32 {
        self.suggested_score
    }
}

/// Result of the G2J subsection (One breakpoint is within an esbalished HI gene,
/// patient's phenotype is either inconsistent with what is expected for LOF of
/// that gene OR unknown).
#[derive(Debug, Clone, PartialEq, Default, serde::Deserialize, serde::Serialize)]
pub struct G2J {
    /// Suggested score for the subsection.
    pub suggested_score: f32,
    /// HI genes overlapping with the CNV.
    pub hi_genes: Vec<GeneIdInfo>,
}

impl SuggestedScore for G2J {
    fn suggested_score(&self) -> f32 {
        self.suggested_score
    }
}

/// Result of the G2L subsection (one or both breakpoints are within gene(s) of
/// no established linical significance).
#[derive(Debug, Clone, PartialEq, Default, serde::Deserialize, serde::Serialize)]
pub struct G2L {
    /// Suggested score for the subsection.
    pub suggested_score: f32,
    /// Genes overlapping with the DUP.
    pub genes: Vec<GeneIdInfo>,
}

impl SuggestedScore for G2L {
    fn suggested_score(&self) -> f32 {
        self.suggested_score
    }
}

/// Enumeration of the categories for the structural variant evaluation, Section 2.
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
pub enum G3 {
    /// <=34 genes.
    G3A(G3Count),
    /// 35-49 genes.
    G3B(G3Count),
    /// >=50 genes.
    G3C(G3Count),
}

/// Result of the 3A subsection (Number of protein-coding RefSeq genes wholly or partially included
/// in the copy-number loss).
#[derive(Debug, Clone, PartialEq, Default, serde::Deserialize, serde::Serialize)]
pub struct G3Count {
    /// Suggested score for the subsection.
    pub suggested_score: f32,
    /// Number of protein-coding RefSeq genes wholly or partially included in the copy-number loss.
    pub num_genes: usize,
    /// The overlapping genes.
    pub genes: Vec<GeneOverlap>,
}

impl SuggestedScore for G3Count {
    fn suggested_score(&self) -> f32 {
        self.suggested_score
    }
}

/// Enumeration of the categories for the structural variant evaluation, Section 4.
///
/// Only 4O can be automatically determined.
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
pub enum G4 {
    /// Overlap with pathogenic variants; must be evaluated by a human.
    ///
    /// This could be one of 4A, 4B, 4C, 4D, 4E, 4L, 4M.
    G4Patho(G4Patho),
    /// Overlap with benign variants; must be evaluated by a human.
    ///
    /// This is roughly equiavalent to 4N.
    G4N(G4N),
    /// Case–control and population evidence; Overlap with common population variation.
    G4O(G4O),
}

/// Result of the 4O subsection (overlap with common population variation).
#[derive(Debug, Clone, PartialEq, Default, serde::Deserialize, serde::Serialize)]
pub struct G4Patho {
    /// Overlapping ClinVar variants, descendingly by overlap.
    pub overlaps: Vec<ClinvarSvOverlap>,
}

impl SuggestedScore for G4Patho {
    fn suggested_score(&self) -> f32 {
        0.0
    }
}

/// Benign variants from ClinVar.
///
/// This is roughly 4N.
#[derive(Debug, Clone, PartialEq, Default, serde::Deserialize, serde::Serialize)]
pub struct G4N {
    /// Overlapping ClinVar variants, descendingly by overlap.
    pub overlaps: Vec<ClinvarSvOverlap>,
}

impl SuggestedScore for G4N {
    fn suggested_score(&self) -> f32 {
        -0.9
    }
}

/// Result of the 4O subsection (overlap with common population variation).
#[derive(Debug, Clone, PartialEq, Default, serde::Deserialize, serde::Serialize)]
pub struct G4O {
    /// Suggested score for the subsection.
    pub suggested_score: f32,
    /// Accession identifiers of overlapping common variants.
    pub overlaps: Vec<GnomadSvOverlap>,
}

impl SuggestedScore for G4O {
    fn suggested_score(&self) -> f32 {
        self.suggested_score
    }
}
