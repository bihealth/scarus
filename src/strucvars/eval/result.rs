//! Common code for evaluation results of CNVs.

use annonars::gnomad_sv::cli::query::Record as GnomadSvRecord;
use annonars::pbs::clinvar::sv::Record as ClinvarSvRecord;
use intervals_general::bound_pair::BoundPair;

use super::{del, dup};

/// Overall result.
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
#[serde(rename_all = "snake_case")]
pub enum Evaluation {
    /// Result of assessing DUPs.
    Dup(dup::result::Evaluation),
    /// Result of assessing DELs.
    Del(del::result::Evaluation),
}

/// Overal assessment of a CNV.
#[derive(
    Debug,
    Clone,
    Copy,
    Default,
    PartialEq,
    strum_macros::EnumIter,
    serde::Deserialize,
    serde::Serialize,
)]
#[serde(rename_all = "snake_case")]
pub enum ClinicalSignificance {
    /// Benign
    Benign,
    /// Likely benign
    LikelyBenign,
    /// Uncertain_significance
    #[default]
    UncertainSignificance,
    /// Likely_pathogenic
    LikelyPathogenic,
    /// Pathogenic
    Pathogenic,
}

/// Score interval to use.
pub type ScoreRange = intervals_general::interval::Interval<f32>;

impl ClinicalSignificance {
    /// Return the scores,
    pub fn range(self) -> ScoreRange {
        match self {
            ClinicalSignificance::Benign => ScoreRange::UnboundedClosedRight { right: 0.99 },
            ClinicalSignificance::LikelyBenign => ScoreRange::LeftHalfOpen {
                bound_pair: BoundPair::new(-0.99, -0.90).expect("invalid interval?"),
            },
            ClinicalSignificance::UncertainSignificance => ScoreRange::Open {
                bound_pair: BoundPair::new(-0.90, 0.90).expect("invalid interval?"),
            },
            ClinicalSignificance::LikelyPathogenic => ScoreRange::LeftHalfOpen {
                bound_pair: BoundPair::new(0.90, 0.99).expect("invalid interval?"),
            },
            ClinicalSignificance::Pathogenic => ScoreRange::UnboundedClosedLeft { left: 0.99 },
        }
    }
}

/// Enumeration describing the PVS1 results.
#[derive(
    Debug,
    Clone,
    Copy,
    PartialEq,
    Eq,
    PartialOrd,
    Ord,
    Default,
    serde::Deserialize,
    serde::Serialize,
)]
#[serde(rename_all = "snake_case")]
pub enum Pvs1Result {
    /// PVS1
    #[default]
    Pvs1,
    /// PVS1_Strong
    Pvs1Strong,
    /// PVS1_Moderate
    Pvs1Moderate,
    /// PVS1_Supporting
    Pvs1Supporting,
}

/// Information about one overlapping ClinVar records.
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
pub struct ClinvarSvOverlap {
    /// Reciprocal overlap.
    pub overlap: f32,
    /// Overlapping ClinVar record.
    pub record: ClinvarSvRecord,
}

/// Information about one overlapping gnomAD records.
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
pub struct GnomadSvOverlap {
    /// Reciprocal overlap.
    pub overlap: f32,
    /// Overlapping ClinVar record.
    pub record: GnomadSvRecord,
}
