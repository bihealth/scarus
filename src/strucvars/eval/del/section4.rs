//! Implementation of evaluation of copy number loss section 4.
//!
//! Note that we can only determine 4O reliably in an automated fashion.  We support
//! reporting overlapping variants from ClinVar but leave them as "dangle" because a
//! human must evaluate the phenotype.

use super::result::Section;
use crate::strucvars::ds::StructuralVariant;

/// Evaluation of deletions, loss of copy number.
///
/// This is mainly used to encapsulate the functionality.  Creating new such
/// objects is very straightforward and cheap.
pub struct Evaluator<'a> {
    /// The parent evaluator.
    #[allow(dead_code)] // TODO
    parent: &'a super::super::Evaluator,
}

impl<'a> Evaluator<'a> {
    /// Create a new `Evaluator`.
    pub fn with_parent(parent: &'a super::super::Evaluator) -> Self {
        Self { parent }
    }

    /// Perform the evaluation of copy number loss Section 3 and all subsection.
    ///
    /// # Arguments
    ///
    /// * `strucvar` - Structural variant to be evaluated.
    ///
    /// # Returns
    ///
    /// The evaluation result.
    ///
    /// # Errors
    ///
    /// If anything goes wrong, it returns a generic `anyhow::Error`.
    pub fn evaluate(&self, strucvar: &StructuralVariant) -> Result<Vec<Section>, anyhow::Error> {
        let _ = strucvar;
        tracing::warn!("Section 4 evaluation not implemented yet");
        Ok(Vec::new())
    }
}

#[cfg(test)]
mod test {
    use crate::strucvars::ds::StructuralVariant;

    use super::super::super::test::global_evaluator_37;

    // #[tracing_test::traced_test]
    #[rstest::rstest]
    fn test_evaluate(
        global_evaluator_37: &super::super::super::Evaluator,
    ) -> Result<(), anyhow::Error> {
        let evaluator = super::Evaluator::with_parent(&global_evaluator_37);

        // TODO: write test after we have an actual implementation
        assert_eq!(
            evaluator.evaluate(&StructuralVariant::default())?,
            Vec::new()
        );

        Ok(())
    }
}
