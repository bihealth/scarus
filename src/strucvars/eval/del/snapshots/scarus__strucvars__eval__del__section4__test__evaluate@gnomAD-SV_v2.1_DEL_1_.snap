---
source: src/strucvars/eval/del/section4.rs
expression: evaluator.evaluate(&strucvar)?
---
- l4:
    l4o:
      suggested_score: -1
      overlaps:
        - overlap: 1
          record:
            gnomad_sv2:
              chrom: "1"
              pos: 21000
              end: 26000
              id: gnomAD-SV_v2.1_DEL_1_1
              filters:
                - FILTER_LOW_CALL_RATE
                - FILTER_UNSTABLE_AF_PCRMINUS
              svType: SV_TYPE_DEL
              alleleCounts:
                - bySex:
                    overall:
                      ac: 138
                      an: 8586
                      af: 0.016073
                      nBiGenos: 4293
                      nHomref: 4155
                      nHet: 138
                      freqHomref: 0.967855
                      freqHet: 0.0321454
                    xx:
                      ac: 68
                      an: 4272
                      af: 0.015918
                      nBiGenos: 2136
                      nHomref: 2068
                      nHet: 68
                      freqHomref: 0.968165
                      freqHet: 0.0318352
                    xy:
                      ac: 69
                      an: 4302
                      af: 0.016039
                      nBiGenos: 2151
                      nHomref: 2082
                      nHet: 69
                      freqHomref: 0.967922
                      freqHet: 0.0320781
                  byPopulation:
                    - population: POPULATION_AFR
                      counts:
                        overall:
                          ac: 27
                          an: 3718
                          af: 0.007262
                          nBiGenos: 1859
                          nHomref: 1832
                          nHet: 27
                          freqHomref: 0.985476
                          freqHet: 0.0145239
                        xx:
                          ac: 11
                          an: 1702
                          af: 0.006463
                          nBiGenos: 851
                          nHomref: 840
                          nHet: 11
                          freqHomref: 0.987074
                          freqHet: 0.012926
                        xy:
                          ac: 16
                          an: 2016
                          af: 0.007937
                          nBiGenos: 1008
                          nHomref: 992
                          nHet: 16
                          freqHomref: 0.984127
                          freqHet: 0.015873
                    - population: POPULATION_AMR
                      counts:
                        overall:
                          ac: 8
                          an: 684
                          af: 0.011696
                          nBiGenos: 342
                          nHomref: 334
                          nHet: 8
                          freqHomref: 0.976608
                          freqHet: 0.0233918
                        xx:
                          ac: 6
                          an: 358
                          af: 0.01676
                          nBiGenos: 179
                          nHomref: 173
                          nHet: 6
                          freqHomref: 0.96648
                          freqHet: 0.0335196
                        xy:
                          ac: 2
                          an: 326
                          af: 0.006135
                          nBiGenos: 163
                          nHomref: 161
                          nHet: 2
                          freqHomref: 0.98773
                          freqHet: 0.0122699
                    - population: POPULATION_EAS
                      counts:
                        overall:
                          ac: 54
                          an: 750
                          af: 0.072
                          nBiGenos: 375
                          nHomref: 321
                          nHet: 54
                          freqHomref: 0.856
                          freqHet: 0.144
                        xx:
                          ac: 25
                          an: 328
                          af: 0.07622
                          nBiGenos: 164
                          nHomref: 139
                          nHet: 25
                          freqHomref: 0.847561
                          freqHet: 0.152439
                        xy:
                          ac: 28
                          an: 418
                          af: 0.066986
                          nBiGenos: 209
                          nHomref: 181
                          nHet: 28
                          freqHomref: 0.866029
                          freqHet: 0.133971
                    - population: POPULATION_EUR
                      counts:
                        overall:
                          ac: 48
                          an: 3346
                          af: 0.014345
                          nBiGenos: 1673
                          nHomref: 1625
                          nHet: 48
                          freqHomref: 0.971309
                          freqHet: 0.028691
                        xx:
                          ac: 26
                          an: 1840
                          af: 0.01413
                          nBiGenos: 920
                          nHomref: 894
                          nHet: 26
                          freqHomref: 0.971739
                          freqHet: 0.0282609
                        xy:
                          ac: 22
                          an: 1498
                          af: 0.014686
                          nBiGenos: 749
                          nHomref: 727
                          nHet: 22
                          freqHomref: 0.970627
                          freqHet: 0.0293725
                    - population: POPULATION_OTHER
                      counts:
                        overall:
                          ac: 1
                          an: 88
                          af: 0.011364
                          nBiGenos: 44
                          nHomref: 43
                          nHet: 1
                          freqHomref: 0.977273
                          freqHet: 0.0227273
                        xx:
                          an: 44
                          nBiGenos: 22
                          nHomref: 22
                          freqHomref: 1
                        xy:
                          ac: 1
                          an: 44
                          af: 0.022727
                          nBiGenos: 22
                          nHomref: 21
                          nHet: 1
                          freqHomref: 0.954545
                          freqHet: 0.0454545
                - cohort: controls
                  bySex:
                    overall:
                      ac: 72
                      an: 4272
                      af: 0.016854
                      nBiGenos: 2136
                      nHomref: 2064
                      nHet: 72
                      freqHomref: 0.966292
                      freqHet: 0.0337079
                    xx:
                      ac: 40
                      an: 2178
                      af: 0.018365
                      nBiGenos: 1089
                      nHomref: 1049
                      nHet: 40
                      freqHomref: 0.963269
                      freqHet: 0.0367309
                    xy:
                      ac: 31
                      an: 2088
                      af: 0.014847
                      nBiGenos: 1044
                      nHomref: 1013
                      nHet: 31
                      freqHomref: 0.970307
                      freqHet: 0.0296935
                  byPopulation:
                    - population: POPULATION_AFR
                      counts:
                        overall:
                          ac: 9
                          an: 1592
                          af: 0.005653
                          nBiGenos: 796
                          nHomref: 787
                          nHet: 9
                          freqHomref: 0.988693
                          freqHet: 0.0113065
                        xx:
                          ac: 6
                          an: 826
                          af: 0.007264
                          nBiGenos: 413
                          nHomref: 407
                          nHet: 6
                          freqHomref: 0.985472
                          freqHet: 0.0145278
                        xy:
                          ac: 3
                          an: 766
                          af: 0.003916
                          nBiGenos: 383
                          nHomref: 380
                          nHet: 3
                          freqHomref: 0.992167
                          freqHet: 0.0078329
                    - population: POPULATION_AMR
                      counts:
                        overall:
                          ac: 4
                          an: 484
                          af: 0.008264
                          nBiGenos: 242
                          nHomref: 238
                          nHet: 4
                          freqHomref: 0.983471
                          freqHet: 0.0165289
                        xx:
                          ac: 3
                          an: 246
                          af: 0.012195
                          nBiGenos: 123
                          nHomref: 120
                          nHet: 3
                          freqHomref: 0.97561
                          freqHet: 0.0243902
                        xy:
                          ac: 1
                          an: 238
                          af: 0.004202
                          nBiGenos: 119
                          nHomref: 118
                          nHet: 1
                          freqHomref: 0.991597
                          freqHet: 0.00840336
                    - population: POPULATION_EAS
                      counts:
                        overall:
                          ac: 39
                          an: 568
                          af: 0.068662
                          nBiGenos: 284
                          nHomref: 245
                          nHet: 39
                          freqHomref: 0.862676
                          freqHet: 0.137324
                        xx:
                          ac: 20
                          an: 268
                          af: 0.074627
                          nBiGenos: 134
                          nHomref: 114
                          nHet: 20
                          freqHomref: 0.850746
                          freqHet: 0.149254
                        xy:
                          ac: 18
                          an: 296
                          af: 0.060811
                          nBiGenos: 148
                          nHomref: 130
                          nHet: 18
                          freqHomref: 0.878378
                          freqHet: 0.121622
                    - population: POPULATION_EUR
                      counts:
                        overall:
                          ac: 20
                          an: 1580
                          af: 0.012658
                          nBiGenos: 790
                          nHomref: 770
                          nHet: 20
                          freqHomref: 0.974684
                          freqHet: 0.0253165
                        xx:
                          ac: 11
                          an: 814
                          af: 0.013514
                          nBiGenos: 407
                          nHomref: 396
                          nHet: 11
                          freqHomref: 0.972973
                          freqHet: 0.027027
                        xy:
                          ac: 9
                          an: 764
                          af: 0.01178
                          nBiGenos: 382
                          nHomref: 373
                          nHet: 9
                          freqHomref: 0.97644
                          freqHet: 0.0235602
                    - population: POPULATION_OTHER
                      counts:
                        overall:
                          an: 48
                          nBiGenos: 24
                          nHomref: 24
                          freqHomref: 1
                        xx:
                          an: 24
                          nBiGenos: 12
                          nHomref: 12
                          freqHomref: 1
                        xy:
                          an: 24
                          nBiGenos: 12
                          nHomref: 12
                          freqHomref: 1
                - cohort: non_neuro
                  bySex:
                    overall:
                      ac: 118
                      an: 6838
                      af: 0.017257
                      nBiGenos: 3419
                      nHomref: 3301
                      nHet: 118
                      freqHomref: 0.965487
                      freqHet: 0.034513
                    xx:
                      ac: 63
                      an: 3640
                      af: 0.017308
                      nBiGenos: 1820
                      nHomref: 1757
                      nHet: 63
                      freqHomref: 0.965385
                      freqHet: 0.0346154
                    xy:
                      ac: 54
                      an: 3186
                      af: 0.016949
                      nBiGenos: 1593
                      nHomref: 1539
                      nHet: 54
                      freqHomref: 0.966102
                      freqHet: 0.0338983
                  byPopulation:
                    - population: POPULATION_AFR
                      counts:
                        overall:
                          ac: 9
                          an: 2032
                          af: 0.004429
                          nBiGenos: 1016
                          nHomref: 1007
                          nHet: 9
                          freqHomref: 0.991142
                          freqHet: 0.00885827
                        xx:
                          ac: 6
                          an: 1100
                          af: 0.005455
                          nBiGenos: 550
                          nHomref: 544
                          nHet: 6
                          freqHomref: 0.989091
                          freqHet: 0.0109091
                        xy:
                          ac: 3
                          an: 932
                          af: 0.003219
                          nBiGenos: 466
                          nHomref: 463
                          nHet: 3
                          freqHomref: 0.993562
                          freqHet: 0.00643777
                    - population: POPULATION_AMR
                      counts:
                        overall:
                          ac: 8
                          an: 684
                          af: 0.011696
                          nBiGenos: 342
                          nHomref: 334
                          nHet: 8
                          freqHomref: 0.976608
                          freqHet: 0.0233918
                        xx:
                          ac: 6
                          an: 358
                          af: 0.01676
                          nBiGenos: 179
                          nHomref: 173
                          nHet: 6
                          freqHomref: 0.96648
                          freqHet: 0.0335196
                        xy:
                          ac: 2
                          an: 326
                          af: 0.006135
                          nBiGenos: 163
                          nHomref: 161
                          nHet: 2
                          freqHomref: 0.98773
                          freqHet: 0.0122699
                    - population: POPULATION_EAS
                      counts:
                        overall:
                          ac: 54
                          an: 750
                          af: 0.072
                          nBiGenos: 375
                          nHomref: 321
                          nHet: 54
                          freqHomref: 0.856
                          freqHet: 0.144
                        xx:
                          ac: 25
                          an: 328
                          af: 0.07622
                          nBiGenos: 164
                          nHomref: 139
                          nHet: 25
                          freqHomref: 0.847561
                          freqHet: 0.152439
                        xy:
                          ac: 28
                          an: 418
                          af: 0.066986
                          nBiGenos: 209
                          nHomref: 181
                          nHet: 28
                          freqHomref: 0.866029
                          freqHet: 0.133971
                    - population: POPULATION_EUR
                      counts:
                        overall:
                          ac: 46
                          an: 3284
                          af: 0.014007
                          nBiGenos: 1642
                          nHomref: 1596
                          nHet: 46
                          freqHomref: 0.971985
                          freqHet: 0.0280146
                        xx:
                          ac: 26
                          an: 1810
                          af: 0.014365
                          nBiGenos: 905
                          nHomref: 879
                          nHet: 26
                          freqHomref: 0.971271
                          freqHet: 0.0287293
                        xy:
                          ac: 20
                          an: 1466
                          af: 0.013643
                          nBiGenos: 733
                          nHomref: 713
                          nHet: 20
                          freqHomref: 0.972715
                          freqHet: 0.0272851
                    - population: POPULATION_OTHER
                      counts:
                        overall:
                          ac: 1
                          an: 88
                          af: 0.011364
                          nBiGenos: 44
                          nHomref: 43
                          nHet: 1
                          freqHomref: 0.977273
                          freqHet: 0.0227273
                        xx:
                          an: 44
                          nBiGenos: 22
                          nHomref: 22
                          freqHomref: 1
                        xy:
                          ac: 1
                          an: 44
                          af: 0.022727
                          nBiGenos: 22
                          nHomref: 21
                          nHet: 1
                          freqHomref: 0.954545
                          freqHet: 0.0454545

